### PR TITLE
Rewrite of AlgoSeekOptionsConverter

### PR DESCRIPTION
**Pull request by AnObfuscator**

### CODE REVIEW

**Comment by jaredbroad** [comment]
The TravisCI is failing with the error:
```
CSC: error CS2001: Source file `AlgoSeekOptionsConverter/TestMethods.cs' could not be found
```
Can you please add this to the repo and store it in the /Tests/ToolBox/AlgoSeekOptionsConverter/ folder?

**Comment by AnObfuscator** [comment]
Yeah, I saw that. For some reason the file wasn't removed from the csproj file when I deleted it. I will fix and push.

File: ToolBox/AlgoSeekOptionsConverter/Program.cs [hunk 1]
```diff
@@ -29,72 +32,32 @@ public class Program
...
-            var directory = "F:/Downloads/AlgoSeek/smaller";
-            directory = "F:/AlgoSeek/20151224";
-            var dataDirectory = "./Data";
+			Environment.SetEnvironmentVariable("MONO_MANAGED_WATCHER", "disabled");
+
+			var sourceDirectory = "/cache/TestData/RawGoog/20151224";
+			var destinationDirectory = "/cache/TestData/Converted";
+
+//			var sourceDirectory = "/cache/DevData/RawDataSmall/20140516";
+//			var destinationDirectory = "/cache/DevData/ConvertedData";
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: Please remove commented code & set the source and destination directory to come from config.

File: ToolBox/AlgoSeekOptionsConverter/Program.cs [hunk 2]
```diff
@@ -20,6 +20,9 @@
 using System.Linq;
 using System.Threading.Tasks;
 using QuantConnect.Logging;
+using  System.Diagnostics;
+using QuantConnect.Data.Market;
+using QuantConnect.Data;
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: There's lots of unused imports; please trim

File: ToolBox/AlgoSeekOptionsConverter/Program.cs [hunk 3]
```diff
@@ -29,72 +32,32 @@ public class Program
...
-                select new DirectoryInfo(dir).FullName;
+			var flushInterval = 1000000L;
 
-            var coarse =
-                from res in new[] {Resolution.Hour, Resolution.Daily}
-                let path = Path.Combine(root, res.ToLower())
-                from dir in Directory.EnumerateDirectories(path)
-                select new DirectoryInfo(dir).FullName;
+			Stopwatch sw = Stopwatch.StartNew();
+			AlgoSeekOptionsConverter.ConvertToFineResolution(sourceDirectory, destinationDirectory, flushInterval);
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: Please convert to be dynamic instantiated type instead of static.

File: ToolBox/AlgoSeekOptionsConverter/Program.cs [hunk 4]
```diff
@@ -29,72 +32,32 @@ public class Program
...
-                    // zip the contents of the directory and then delete the directory
-                    Compression.ZipDirectory(dir, dir + ".zip", false);
-                    Directory.Delete(dir, true);
-                }
-                catch (Exception err)
-                {
-                    Log.Error(err, "Zipping " + dir);
-                }
-            });
+			Log.Trace(String.Format("Conversion finished in time: {0}", sw.Elapsed));
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: Please restore the compression logic. This is needed to package up the csv files, I see a method linked in the other class but not called.

Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: Please also use "var" instead of any type declarations and lower case object names ("string" not "String")

File: ToolBox/AlgoSeekOptionsConverter/Program.cs (hunk 3 above)
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Program.cs]: var stopwatch, no contractions where possible. Datetime would also suffice since its running for a long time.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+using QuantConnect.Data.Market;
+using System.Linq;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace QuantConnect.ToolBox.AlgoSeekOptionsConverter
+{
+	public class AlgoSeekOptionsConverter
+	{
+		public static void ConvertToFineResolution(string sourceDirectory, string destinationDirectory, long flushInterval)
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: Please put summary comments on all public methods.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+		{
+			var quoteProcessor = new InMemoryDataProcessor(TickType.Quote, x => ((Tick) x).TickType == TickType.Quote);
+			var tradeProcessor = new InMemoryDataProcessor(TickType.Trade, x => ((Tick) x).TickType == TickType.Trade);
+
+			var files = Directory.EnumerateFiles(sourceDirectory).OrderByDescending(x => new FileInfo(x).Length);
+			var optionsReaders = files.Select(file => new AlgoSeekOptionsReader(file)).ToList();
+
+			var parsingTime = new Stopwatch();
+			var processingTime = new Stopwatch();
+			var diskTime = new Stopwatch();
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: Please remove all bench-marking assists; at this stage they gum up the code. Simple logging of progress every 5-10 sec should be enough for production code.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+
+			var totalLinesProcessed = 0L;
+			while(optionsReaders.Any(reader => reader.HasNext))
+			{
+				parsingTime.Start();
+				var activeReaders = optionsReaders.Where(reader => reader.HasNext);
+				var nextTick = activeReaders.EarliestTick().Take();
+				var dataTime = nextTick.Time;
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: That's a cool implementation :)

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+				quoteProcessor.Process(nextTick);
+				tradeProcessor.Process(nextTick);
+				processingTime.Stop();
+
+				totalLinesProcessed++;
+				if (totalLinesProcessed % flushInterval == 0)
+				{
+					var cutoffTime = new DateTime(dataTime.Year, dataTime.Month, dataTime.Day, dataTime.Hour, dataTime.Minute, 0, 0) - TimeSpan.FromMinutes(1);
+
+					Log.Trace("Processed {0,3}M lines; Data Time is: {1}; Memory in use: {2} MB", totalLinesProcessed/flushInterval, dataTime, Info.CurrentMemoryUsage);
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: In general we don't like statics like Info in Lean, please just make it a property of the class or private variable of the class.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+				processingTime.Start();
+				quoteProcessor.Process(nextTick);
+				tradeProcessor.Process(nextTick);
+				processingTime.Stop();
+
+				totalLinesProcessed++;
+				if (totalLinesProcessed % flushInterval == 0)
+				{
+					var cutoffTime = new DateTime(dataTime.Year, dataTime.Month, dataTime.Day, dataTime.Hour, dataTime.Minute, 0, 0) - TimeSpan.FromMinutes(1);
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: also close enough is our extension `dataTime.RoundDown(TimeSpan.FromMinutes(1))`

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+						count++;
+						if (count%logInterval == 0)
+						{
+							Log.Trace("({0}): Parsed {1,3}M lines", inputFile, count/logInterval);
+						}
+						const int columns = 11;
+						var csv = line.ToCsv(columns);
+						if (csv.Count < columns) 
+						{
+							continue;
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: When does this happen? Can you error-log those lines and see whats going on?

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+using System.IO;
+using System.Threading;
+using QuantConnect.Data;
+using QuantConnect.Data.Market;
+using QuantConnect.Logging;
+
+
+namespace QuantConnect.ToolBox.AlgoSeekOptionsConverter
+{
+	public class AlgoSeekOptionsReader : IDisposable
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs]: + Please make IEnumerable :) You're 90% of the way there already. 
+ Comments on all public classes / just lots more comments in general.

File: ToolBox/AlgoSeekOptionsConverter/InMemoryDataProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+namespace QuantConnect.ToolBox.AlgoSeekOptionsConverter
+{
+	/*
+	 * Possible Replacement:
+	 * https://gist.github.com/AnObfuscator/a619eea8e9b66ef4967bd2c679400d39
+	 */
+	public class InMemoryDataProcessor : IDataProcessor
+	{
+		private readonly TickType _tickType;
+		private readonly ConcurrentDictionary<Symbol, TickContainer> _ticksBySymbol;
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/InMemoryDataProcessor.cs]: Please make this not a concurrent dictionary - it could be hiding design flaws. There should be no way with a single threaded streaming design one tick after another that it can simultaneously add to the dictionary from two places.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+		private readonly string _file;
+		private readonly Stream _stream;
+		private readonly StreamReader _streamReader;
+		private readonly DateTime _referenceDate;
+
+		private BaseData _nextLine = null;
+
+		public bool HasNext { get { return _nextLine != null; } }
+
+		public BaseData Peek { get { return _nextLine; } }
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs]: Ideally "line" should refer to line from the file. It should be an IEnumerable of BaseData so _nextData etc Having _nextLine return BaseData is confusing.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+
+		public AlgoSeekOptionsReader(string file)
+		{
+			_file = file;
+
+			var streamProvider = StreamProvider.ForExtension(Path.GetExtension(file));
+			_stream = streamProvider.Open(file).First();
+			_streamReader = new StreamReader(_stream);
+
+			_nextLine = GetNextValidLine();
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsReader.cs]: This should be MoveNext() which would set public property Current to follow IEnumerator pattern

File: ToolBox/QuantConnect.ToolBox.csproj [hunk 15]
```diff
@@ -117,14 +117,6 @@
     <Reference Include="Ionic.Zip">
       <HintPath>..\packages\DotNetZip.1.9.3\lib\net20\Ionic.Zip.dll</HintPath>
     </Reference>
-    <Reference Include="Newtonsoft.Json, Version=7.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed, processorArchitecture=MSIL">
```
Comment (jaredbroad) [inline ToolBox/QuantConnect.ToolBox.csproj]: I would leave these in. You don't know who else in the ToolBox project is using them. Same with Noda below.

Comment (jaredbroad) [inline ToolBox/QuantConnect.ToolBox.csproj]: Oops yes see you've upgraded. Unless critical to this specific PR would leave them as they were and submit a separate PR for the upgrade if it makes sense.

File: ToolBox/AlgoSeekOptionsConverter/Utilities.cs [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+using QuantConnect.Logging;
+using QuantConnect.Data;
+using QuantConnect.Util;
+
+
+namespace QuantConnect.ToolBox.AlgoSeekOptionsConverter
+{
+	public static class Extensions
+	{
+		public static AlgoSeekOptionsReader EarliestTick(this IEnumerable<AlgoSeekOptionsReader> readers)
```
Comment (jaredbroad) [inline ToolBox/AlgoSeekOptionsConverter/Utilities.cs]: This whole Utilties.cs can be removed really. 
+ The EarliestTick is too specific // only used in 1 place.
+ When you remove the benchmark logging the Info is only used in 1 place as well,
+ Resolutions is used a few places but really should be solved in a loop: so points to a design flaw
```
     foreach(var resolution in X) process.AddConsolidators(new Consolidator(resolution));
```

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs (hunk 10 above)
Comment (AnObfuscator) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: I don't know. This was lifted from Michael's implementation.

File: ToolBox/AlgoSeekOptionsConverter/InMemoryDataProcessor.cs (hunk 12 above)
Comment (AnObfuscator) [inline ToolBox/AlgoSeekOptionsConverter/InMemoryDataProcessor.cs]: That's fine. This class originally did get data from multiple threads, hence the thread-safety, but that implementation had to be abandoned.

File: ToolBox/QuantConnect.ToolBox.csproj (hunk 15 above)
Comment (AnObfuscator) [inline ToolBox/QuantConnect.ToolBox.csproj]: I didn't mean to; MonoDevelop did that for me accidentally. I didn't even notice.

File: ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs (hunk 6 above)
Comment (AnObfuscator) [inline ToolBox/AlgoSeekOptionsConverter/AlgoSeekOptionsConverter.cs]: Would you still like each file to record it's progress, or just overall progress? I wasn't sure what we'd want here.

**Comment by AnObfuscator** [comment]
I couldn't agree more, @mchandschuh. I ordinarily follow a similar commit message pattern (like [here](https://github.com/AnObfuscator/pyWireMock/commits/master) ), but I got in a rush to get this code out and dropped the ball here.

I will absolutely be more careful to commit in smaller chunks, and use better messages, in the future.

**Comment by jaredbroad** [comment]
Closing as rewritten in the futures branch
